# HG changeset patch
# User Betterbird <betterbird@betterbird.eu>
# Date 1765299151 -3600
# Parent  ddaf0d561f511e535a25c68f83a9e177800896e3
Feature: CardDAV Categories: Add missing parts: List view, search, etc.

diff --git a/mail/components/addrbook/content/abSearchDialog.js b/mail/components/addrbook/content/abSearchDialog.js
--- a/mail/components/addrbook/content/abSearchDialog.js
+++ b/mail/components/addrbook/content/abSearchDialog.js
@@ -118,8 +118,17 @@ function searchOnLoad() {
       },
       hidden: true,
     },
     {
+      id: "Categories",
+      l10n: {
+        header: "about-addressbook-column-header-categories2",
+        menuitem: "about-addressbook-column-label-categories2",
+        cell: "about-addressbook-cell-categories2",
+      },
+      hidden: true,
+    },
+    {
       id: "PhoneNumbers",
       l10n: {
         header: "about-addressbook-column-header-phonenumbers2",
         menuitem: "about-addressbook-column-label-phonenumbers2",
@@ -338,8 +347,11 @@ function onSearch() {
         break;
       case Ci.nsMsgSearchAttrib.Nickname:
         attrs = ["NickName"];
         break;
+      case Ci.nsMsgSearchAttrib.Categories:
+        attrs = ["Categories"];
+        break;
       case Ci.nsMsgSearchAttrib.WorkPhone:
         attrs = ["WorkPhone"];
         break;
       case Ci.nsMsgSearchAttrib.HomePhone:
diff --git a/mail/components/addrbook/content/aboutAddressBook.js b/mail/components/addrbook/content/aboutAddressBook.js
--- a/mail/components/addrbook/content/aboutAddressBook.js
+++ b/mail/components/addrbook/content/aboutAddressBook.js
@@ -1379,8 +1379,16 @@ var cardsPane = {
       },
       hidden: true,
     },
     {
+      id: "Categories",
+      l10n: {
+        header: "about-addressbook-column-header-categories2",
+        menuitem: "about-addressbook-column-label-categories2",
+      },
+      hidden: true,
+    },
+    {
       id: "PhoneNumbers",
       l10n: {
         header: "about-addressbook-column-header-phonenumbers2",
         menuitem: "about-addressbook-column-label-phonenumbers2",
diff --git a/mail/components/addrbook/content/modules/AddrBookDataAdapter.mjs b/mail/components/addrbook/content/modules/AddrBookDataAdapter.mjs
--- a/mail/components/addrbook/content/modules/AddrBookDataAdapter.mjs
+++ b/mail/components/addrbook/content/modules/AddrBookDataAdapter.mjs
@@ -455,8 +455,20 @@ class AddrBookDataRow extends TreeDataRo
           return AddrBookDataRow.listFormatter.format(
             addresses.filter(Boolean)
           );
         }
+        case "Categories": {
+          let value;
+          if (supportsVCard) {
+            value = vCardProperties.getFirstValue("categories");
+          } else {
+            value = getProperty("Categories", "");
+          }
+          if (typeof(value) === "object") {
+            value = value.join(", ");
+          }
+          return value;
+        }
         case "JobTitle":
         case "Title":
           if (supportsVCard) {
             return vCardProperties.getFirstValue("title");
diff --git a/mail/locales/en-US/chrome/messenger/search-attributes.properties b/mail/locales/en-US/chrome/messenger/search-attributes.properties
--- a/mail/locales/en-US/chrome/messenger/search-attributes.properties
+++ b/mail/locales/en-US/chrome/messenger/search-attributes.properties
@@ -34,8 +34,9 @@ City=City
 Street=Street
 Title=Title
 Organization=Organization
 Department=Department
+Categories=Categories
 # more mailnews
 FromToCcOrBcc=From, To, Cc or Bcc
 JunkScoreOrigin=Junk Score Origin
 JunkPercent=Junk Percent
diff --git a/mail/locales/en-US/messenger/addressbook/aboutAddressBook.ftl b/mail/locales/en-US/messenger/addressbook/aboutAddressBook.ftl
--- a/mail/locales/en-US/messenger/addressbook/aboutAddressBook.ftl
+++ b/mail/locales/en-US/messenger/addressbook/aboutAddressBook.ftl
@@ -131,8 +131,18 @@ about-addressbook-column-label-nickname2
 about-addressbook-cell-nickname2 =
   .aria-label = Nickname
   .title = { $title }
 
+about-addressbook-column-header-categories2 = Categories
+  .title = Sort by categories
+about-addressbook-column-label-categories2 =
+  .label = Categories
+# Variables:
+# $title (String) - Contact categories for tooltip.
+about-addressbook-cell-categories2 =
+  .aria-label = Categories
+  .title = { $title }
+
 about-addressbook-column-header-phonenumbers2 = Phone Numbers
   .title = Sort by phone numbers
 about-addressbook-column-label-phonenumbers2 =
   .label = Phone Numbers
diff --git a/mailnews/search/public/nsMsgSearchCore.idl b/mailnews/search/public/nsMsgSearchCore.idl
--- a/mailnews/search/public/nsMsgSearchCore.idl
+++ b/mailnews/search/public/nsMsgSearchCore.idl
@@ -85,8 +85,9 @@ interface nsMsgSearchAttrib : nsISupport
     const nsMsgSearchAttribValue Street = 30;
     const nsMsgSearchAttribValue Title = 31;
     const nsMsgSearchAttribValue Organization = 32;
     const nsMsgSearchAttribValue Department = 33;
+    const nsMsgSearchAttribValue Categories = 34;
 
     // 34 - 43, reserved for ab / LDAP;
     const nsMsgSearchAttribValue HasAttachmentStatus = 44;
     const nsMsgSearchAttribValue JunkStatus = 45;
diff --git a/mailnews/search/src/nsMsgSearchAdapter.cpp b/mailnews/search/src/nsMsgSearchAdapter.cpp
--- a/mailnews/search/src/nsMsgSearchAdapter.cpp
+++ b/mailnews/search/src/nsMsgSearchAdapter.cpp
@@ -894,8 +894,9 @@ static struct {
     {nsMsgSearchAttrib::Street, "Street"},
     {nsMsgSearchAttrib::Title, "Title"},
     {nsMsgSearchAttrib::Organization, "Organization"},
     {nsMsgSearchAttrib::Department, "Department"},
+    {nsMsgSearchAttrib::Categories, "Categories"},
     {nsMsgSearchAttrib::AllAddresses, "FromToCcOrBcc"},
     {nsMsgSearchAttrib::JunkScoreOrigin, "JunkScoreOrigin"},
     {nsMsgSearchAttrib::JunkPercent, "JunkPercent"},
     {nsMsgSearchAttrib::HasAttachmentStatus, "AttachmentStatus"},
@@ -1077,8 +1078,11 @@ nsresult nsMsgSearchValidityManager::Set
 
   rv = EnableDirectoryAttribute(aTable, nsMsgSearchAttrib::Nickname);
   NS_ENSURE_SUCCESS(rv, rv);
 
+  rv = EnableDirectoryAttribute(aTable, nsMsgSearchAttrib::Categories);
+  NS_ENSURE_SUCCESS(rv, rv);
+
   rv = EnableDirectoryAttribute(aTable, nsMsgSearchAttrib::WorkPhone);
   NS_ENSURE_SUCCESS(rv, rv);
 
   rv = EnableDirectoryAttribute(aTable, nsMsgSearchAttrib::HomePhone);
